Description: Strict focus mode (/apps/metacity/general/focus_mode = strict)
Author: LaMont Jones <lamont at debian.org>

Index: mutter-2.91.4/src/core/display.c
===================================================================
--- mutter-2.91.4.orig/src/core/display.c	2010-11-23 08:02:03.000000000 +1100
+++ mutter-2.91.4/src/core/display.c	2010-12-23 10:09:52.846944000 +1100
@@ -2019,6 +2019,7 @@
         {
           switch (meta_prefs_get_focus_mode ())
             {
+            case META_FOCUS_MODE_STRICT:
             case META_FOCUS_MODE_SLOPPY:
             case META_FOCUS_MODE_MOUSE:
               display->mouse_mode = TRUE;
Index: mutter-2.91.4/src/core/prefs.c
===================================================================
--- mutter-2.91.4.orig/src/core/prefs.c	2010-12-03 10:45:59.000000000 +1100
+++ mutter-2.91.4/src/core/prefs.c	2010-12-23 10:09:52.846944000 +1100
@@ -199,6 +199,7 @@
     { META_FOCUS_MODE_CLICK,  "click" },
     { META_FOCUS_MODE_SLOPPY, "sloppy" },
     { META_FOCUS_MODE_MOUSE,  "mouse" },
+    { META_FOCUS_MODE_STRICT, "strict" },
     { 0, NULL },
   };
 
Index: mutter-2.91.4/src/core/window.c
===================================================================
--- mutter-2.91.4.orig/src/core/window.c	2010-12-14 02:25:53.000000000 +1100
+++ mutter-2.91.4/src/core/window.c	2010-12-23 10:09:52.856944000 +1100
@@ -2708,7 +2708,7 @@
   if (window->showing_for_first_time)
     {
       window->showing_for_first_time = FALSE;
-      if (takes_focus_on_map)
+      if (takes_focus_on_map && meta_prefs_get_focus_mode () != META_FOCUS_MODE_STRICT)
         {
           guint32     timestamp;
 
Index: mutter-2.91.4/src/core/workspace.c
===================================================================
--- mutter-2.91.4.orig/src/core/workspace.c	2010-11-30 08:04:46.000000000 +1100
+++ mutter-2.91.4/src/core/workspace.c	2010-12-23 10:09:52.856944000 +1100
@@ -1267,7 +1267,8 @@
         }
       else if (meta_prefs_get_focus_mode () == META_FOCUS_MODE_SLOPPY)
         focus_ancestor_or_mru_window (workspace, not_this_one, timestamp);
-      else if (meta_prefs_get_focus_mode () == META_FOCUS_MODE_MOUSE)
+      else if (meta_prefs_get_focus_mode () == META_FOCUS_MODE_MOUSE ||
+               meta_prefs_get_focus_mode () == META_FOCUS_MODE_STRICT)
         {
           meta_topic (META_DEBUG_FOCUS,
                       "Setting focus to no_focus_window, since no valid "
Index: mutter-2.91.4/src/include/common.h
===================================================================
--- mutter-2.91.4.orig/src/include/common.h	2010-11-20 06:39:16.000000000 +1100
+++ mutter-2.91.4/src/include/common.h	2010-12-23 10:09:52.856944000 +1100
@@ -165,7 +165,8 @@
 {
   META_FOCUS_MODE_CLICK,
   META_FOCUS_MODE_SLOPPY,
-  META_FOCUS_MODE_MOUSE
+  META_FOCUS_MODE_MOUSE,
+  META_FOCUS_MODE_STRICT
 } MetaFocusMode;
 
 typedef enum
